src/js/controller/qoe.ts

```diff
@@ -7,15 +7,43 @@ import {
     MEDIA_VISUAL_QUALITY
 } from 'events/events';
 import Timer from 'api/timer';
+import type {
+    TimeEvent,
+    InternalPlayerState
+} from 'events/events';
+import type Model from 'controller/model';
+import type { MediaModel } from 'controller/model';
+import type { ProgramController } from 'program/program-controller';
 
 const TAB_HIDDEN = 'tabHidden';
 const TAB_VISIBLE = 'tabVisible';
 
+interface QoeModel extends Model {
+    _qoeItem: QoeItem;
+    _triggerFirstFrame: () => void;
+    _onTime: (evt: TimeEvent) => void;
+    _onPlayAttempt: () => void;
+    _onTabVisible?: (modelChanged: Model, activeTab: boolean) => void;
+}
+
+class QoeItem extends Timer {
```

Comment:
refactor really paid off

Reply:
Yes it did. It's like night and day. Great suggestion!